20170908 generate historical diff

mozmark opened the pull request:
The changes to oneCRL.go are obsolete (assuming we merge pr #68), but other than that, this does what Gerv asked for (some tidying may be required).

mozmark commented:
Hold fire on reviewing this until PR #68 is merged

jcjones commented on summarizeOneCRL/main.go:
```diff
@@ -0,0 +1,196 @@
+package main
```
meta-note: we should file an issue to put MPL headers on all these files

mozmark replied:
Filed issue #74

jcjones commented on summarizeOneCRL/main.go:
```diff
@@ -0,0 +1,196 @@
+	// render the report
+	if "HTML" == strings.ToUpper(*typePtr) {
+		renderToHTML(reportLines, *outFilePtr)
+	}
```
should probably print an error if it's an unrecognized type

jcjones commented on summarizeOneCRL/main.go:
```diff
@@ -0,0 +1,196 @@
+
+func main() {
+	urlPtr := flag.String("url", "https://ccadb-public.secure.force.com/mozilla/PublicIntermediateCertsRevokedWithPEMCSV", "the URL of the salesforce data")
+	typePtr := flag.String("type", "html", "the type of report to generate")
```
`(Options: html)`

jcjones commented on summarizeOneCRL/main.go:
```diff
@@ -0,0 +1,196 @@
+}
+
+func renderToHTML(reportLines []ReportLine, filename string) {
+	f, err := os.Create(filename)
```
File open/close should be at the caller, otherwise all the implementations will have to duplicate this

jcjones commented on summarizeOneCRL/main.go:
```diff
@@ -0,0 +1,196 @@
+
+		issuerString := base64.StdEncoding.EncodeToString(cert.RawIssuer)
+		serialBytes, err := asn1.Marshal(cert.SerialNumber)
+
```
nit: unnecessary extra newline

jcjones commented on summarizeOneCRL/main.go:
```diff
@@ -0,0 +1,196 @@
+		serialString := base64.StdEncoding.EncodeToString(serialBytes[2:])
+		
+		if false {
+			fmt.Printf("Look! %s %s\n", issuerString, serialString);
```
I will never look at this line. ;)

(Please remove dead code)

jcjones commented on summarizeOneCRL/main.go:
```diff
@@ -0,0 +1,196 @@
+		revocationInfoToCCADBEntry[oneCRL.StringFromIssuerSerial(issuerString, serialString)] = revoked
+	}
+
+	if err != nil  {
```
What is this `err` checking? I know it's coming out of the loop, but this is not a good design, it's not clear what the return will be. 

If we want to exit early, then let's do it where it matters, not ... out here in the whitespace desert.

mozmark replied:
Nah, we're out of the loop already here. I think what's happened here is that I put the loop in between the call to FetchRevokedCertInfoFrom and the error check.

jcjones commented on summarizeOneCRL/main.go:
```diff
@@ -0,0 +1,196 @@
+
+		reportLine.SerialNumber, _ = oneCRL.SerialToString(entry.SerialNumber, false, false)
+
+		reportLine.IssuerName, _ = oneCRL.DNToRFC4514(entry.IssuerName)
```
Do we really not care about these errors?

jcjones commented on util/bugs.go:
```diff
@@ -115,3 +126,33 @@ func AttachToBug(bugNum int, apiKey string, attachments []Attachment, conf *conf
+			bugNumString = bugNum
+		}
+	}
+	// TODO: we should appropriately escape the bug number and fields
```
yeah, agreed.

jcjones commented on util/bugs.go:
```diff
@@ -115,3 +126,33 @@ func AttachToBug(bugNum int, apiKey string, attachments []Attachment, conf *conf
+		}
+	}
+	// TODO: we should appropriately escape the bug number and fields
+	getUrl := fmt.Sprintf(conf.BugzillaBase + "/rest/bug?id=%s&include_fields=%s", bugNumString, getBugDataIncludeFields)
```
Use the https://golang.org/pkg/net/url/ package :)

mozmark commented on salesforce/salesforce.go:
```diff
@@ -51,25 +54,63 @@ func FetchSalesforceCSV(stream io.ReadCloser) SalesforceCSV {
+	if 0 == strings.Index(strings.ToUpper(location), "HTTP") {
+		if 0 != strings.Index(strings.ToUpper(location), "HTTPS") {
+			// cowardly refuse to get cert info from a non-https URL
+			return make([]RevokedCertInfo, 0), errors.New("Cowardly refusing to load data from a non-HTTPS URL")
```
Filed issue #74

mozmark replied:
Agreed. Filed Issue #76

mozmark commented on oneCRL/oneCRL.go:
```diff
@@ -96,7 +97,7 @@ func getDataFromURL(url string, user string, pass string) ([]byte, error) {
 }
 
-func FetchExistingRevocations(url string) ([]string, error) {
+func FetchExistingRecords(url string) (*Records, error) {
```
This has gone now - PR #68 made the necessary change to FetchExistingRevocations that rendered the creation of FetchExistingRecords unnecessary (thus https://github.com/mozilla/OneCRL-Tools/pull/69#issuecomment-332565644 )

mozmark commented:
@jcjones I messed up my merge and included all of the work addressing feedback to the merge commit (because I suck at git). Hopefully the PR is still review-able?
